src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumWatcher.kt

```diff
@@ -1,597 +1,303 @@
+                                    logger.info { "delaying publication of txid=${tx.txid} until block=$cltvTimeout (curblock=$blockCount)" }
+                                    val block2tx = state.block2tx + (cltvTimeout to state.block2tx.getOrElse(cltvTimeout) { setOf() } + tx)
+                                    state = state.copy(block2tx = block2tx)
+                                }
 
-    private suspend fun checkIsUpToDate() {
-
-        // Get a list of timestamps from the client for all outgoing requests.
-        // This returns an array of `RequestResponseTimestamp` instances, which includes:
-        // - id: Int
-        // - request: ElectrumRequest
-        // - lastResponseTimestamp: Long?
-        //
-        // If lastResponseTimestamp is null, we have an outstanding request pending.
-        //
-        val info = client.requestResponseTimestamps()
-        if (info == null) {
-            // Client isn't connected (i.e. client.state !is ClientRunning)
-            return
-        }
+                                else -> {
+                                    logger.info { "publishing tx=[${tx.txid} / $tx]" }
+                                    client.broadcastTransaction(tx)
+                                    state = state.copy(sent = state.sent + tx)
+                                }
+                            }
+                        }
+                    }
 
-        // We want to emit an "up-to-date" notification to the client.
-        // This is used during WatchTower operation, where the client connects to electrum
-        // in order to ensure the acinq server didn't publish an old commit (i.e. attempt to cheat).
-        //
-        // So we want to emit this message when we're fairly confident the electrum server
-        // has delivered all the pending information we need. Here's how we do it:
-        // - we track the last response time (timestamp in millis) per request
-        // - we filter the list to only include those operations related to the WatchTower
-        // - we wait until all such responses have arrived, and are older than 5 seconds
-
-        val now = currentTimestampMillis()
-        val isUpToDate = info.filter {
-            when (it.request) {
-                is ScriptHashSubscription -> true
-                is GetScriptHashHistory -> true
-                is GetTransaction -> true
-                is GetMerkle -> true
-                is BroadcastTransaction -> true
-                else -> false
+                    is WatcherCommand.NotifyIfReady -> {
+                        if (state.isConnected) {
+                            state.idleSince?.let {
+                                if (it < currentTimestampMillis() - 5000) {
+                                    // no requests in progress and watcher has been idle for more than 5s
+                                    logger.debug { "Watcher is up-to-date and idle for more than 5s" }
```

Comment:
This log is very spammy (every 2sec) and I don't understand the purpose.

Reply:
removed in https://github.com/ACINQ/lightning-kmp/pull/411/commits/f66fd8221e492a36567df03027935e8879691679

Reply:
What is the purpose of `NotifyIfReady`? What are the hardcoded 2s and 5s durations about?

Reply:
It is used to replicate what is done on the master branch to send out `NotifyUpToDateEvent` events. With the new watcher there is no concept of "pending requests": whenever the watcher is ready to process a command or event is means that is has processed all the calls it needs to make to the electrum server, so we don't use request timestamps anymore but instead know when the watcher has been idle for more than X seconds. The hard-coded 2s and 5s duration are copied from the master branch, so the new watcher should behave exactly like the current one.

Reply:
I can't find any reference of `NotifyUpToDateEvent` in the Phoenix codebase. If we are not using it, let's drop it. @dpad85 @robbiehanson please confirm.

Reply:
I could not find any references to `NotifyIfReady` or  `NotifyUpToDateEvent` either on phoenix `master` branch or the iOS production branch. We don't use these events, I think we can drop them.

Reply:
https://github.com/ACINQ/lightning-kmp/pull/411/commits/3e16024d77a450a22b15d13aff5249547caf1cfd removes ` NotifyUpToDateEvent`

Reply:
Phew, feels lighter.

Reply:
[ElectrumWatcher.openUpToDateFlow](https://github.com/ACINQ/lightning-kmp/blob/57f619f3605cc8c7eba1ae1bf29342103c5429c0/src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumWatcher.kt#L429) is being by the WatchTower [here](https://github.com/ACINQ/phoenix/blob/9306dc669bf33a059ccae7984f0e216ce8b1d9f4/phoenix-ios/phoenix-ios/officers/WatchTower.swift#L168) (via a wrapping [publisher](https://github.com/ACINQ/phoenix/blob/9306dc669bf33a059ccae7984f0e216ce8b1d9f4/phoenix-ios/phoenix-ios/kotlin/KotlinPublishers.swift#L287))

Reply:
@robbiehanson can you please test commit https://github.com/ACINQ/lightning-kmp/pull/411/commits/08640a6c61435f96163862c35e1f7c89b8370168 ?

Reply:
:+1: I was able to compile phoenix master branch (with some minor tweaks) against https://github.com/ACINQ/lightning-kmp/commit/08640a6c61435f96163862c35e1f7c89b8370168